src/unix/fs.c

```diff
@@ -358,22 +447,122 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
   return n;
 }
 
+static int uv__fs_opendir(uv_fs_t* req) {
+  DIR *dir_stream;
+  uv__dirent_t *dirent;
+  uv_dir_t *dir;
+  size_t len;
 
-static ssize_t uv__fs_readlink(uv_fs_t* req) {
-  ssize_t len;
-  char* buf;
+  assert(req);
 
-  len = pathconf(req->path, _PC_PATH_MAX);
+  dir_stream = opendir(req->path);
+  if (dir_stream == NULL)
+    return -1;
 
-  if (len == -1) {
-#if defined(PATH_MAX)
-    len = PATH_MAX;
-#else
-    len = 4096;
-#endif
+  len = uv__fs_direntry_size(dir_stream);
+  if (len == (size_t)-1)
+    return -1;
+
+  /*
+  * Allocate the space for each directory entry just once instead of
+  * once per directory entry.
+  */
+  dirent = malloc(len + 1);
+  if (dirent == NULL) {
+   errno = ENOMEM;
+   goto failed;
+  }
+
+  dir = malloc(sizeof(*dir));
+  if (dir == NULL) {
+    errno = ENOMEM;
+    goto failed;
+  }
+
+  dir->dir = dir_stream;
+  dir->dirent = dirent;
+
+  req->dir = dir;
+  req->ptr = (void*)req->dir;
+
+  return 0;
+
+failed:
+  if (errno == ENOMEM) {
+    if (dirent != NULL)
+      free(dirent);
+    if (dir != NULL)
+      free(dir);
+  }
+
+  return -1;
+}
+
+static int uv__fs_readdir(uv_fs_t* req) {
+  uv_dirent_t *dirents;
+  struct dirent *res;
+  int r;
+  unsigned int dirent_idx;
+
+  assert(req);
+
+  assert(req->dir);
+  assert(req->dir->dir);
+  assert(req->dir->dirent);
+
+  assert(req->dirents);
+  assert(req->nentries > 0);
+
+  dirents = req->dirents;
+  req->ptr = req->dirents;
+
+  for (dirent_idx = 0; dirent_idx < req->nentries; ++dirent_idx) {
+    r = readdir_r(req->dir->dir, req->dir->dirent, &res);
```

Comment:
@whitlockjc @saghul 
Hey, I see that `readdir_r` was deprecated in glibc by https://sourceware.org/ml/libc-alpha/2016-02/msg00093.html 
The reason is mostly about indecisive name length, see here - http://man7.org/linux/man-pages/man3/readdir_r.3.html
The suggestion is to use `readdir` which would also eliminate the need for `uv__fs_direntry_size` and its guesstimations.
In multithreaded environment, such as with uv fs workers, calling `readdir` on the same `DIR` stream is racy, and should be protected by mutex per dir.